vercel/resource_firewall_config.go

```diff
@@ -0,0 +1,878 @@
+				ipRules[i].Notes = state.IPRules.Rules[i].Notes
+			}
+		}
+
+		cfg.IPRules = &IPRules{Rules: ipRules}
+	}
+
+	managedRulesets := &FirewallManagedRulesets{}
+	if conf.ManagedRulesets != nil && conf.CRS != nil {
+		cfg.ManagedRulesets = managedRulesets
+		cfg.ManagedRulesets.OWASP = fromCRS(conf.CRS, state.ManagedRulesets)
+	}
+
+	return cfg
+}
+
+func (f *FirewallConfig) toClient() client.FirewallConfig {
+	conf := client.FirewallConfig{
+		ProjectID: f.ProjectID.ValueString(),
+		TeamID:    f.TeamID.ValueString(),
+		Enabled:   f.Enabled.IsNull() || f.Enabled.ValueBool(),
+	}
+
+	if f.ManagedRulesets != nil {
+		conf.ManagedRulesets = make(map[string]client.ManagedRule)
+		if f.ManagedRulesets.OWASP != nil {
+			conf.ManagedRulesets["owasp"] = client.ManagedRule{
+				Active: true,
+			}
+			conf.CRS = make(map[string]client.CoreRuleSet)
+			for key, value := range f.ManagedRulesets.OWASP.ToMap() {
+				if value != nil {
+					conf.CRS[key] = client.CoreRuleSet{
+						Action: value.Action.ValueString(),
+						Active: value.Active.IsNull() || value.Active.ValueBool(),
+					}
+				}
+			}
+		}
+	}
+	if f.Rules != nil && len(f.Rules.Rules) > 0 {
+		for _, rule := range f.Rules.Rules {
+			conf.Rules = append(conf.Rules, client.FirewallRule{
+				ID:             rule.ID.ValueString(),
+				Name:           rule.Name.ValueString(),
+				Description:    rule.Description.ValueString(),
+				Active:         rule.Active.IsNull() || rule.Active.ValueBool(),
+				ConditionGroup: rule.Conditions(),
+				Action: client.Action{
+					Mitigate: rule.Mitigate(),
+				},
+			})
+		}
+	}
+
+	if f.IPRules != nil && len(f.IPRules.Rules) > 0 {
+		for _, iprule := range f.IPRules.Rules {
+			conf.IPRules = append(conf.IPRules, client.IPRule{
+				ID:       iprule.ID.ValueString(),
+				Hostname: iprule.Hostname.ValueString(),
+				IP:       iprule.IP.ValueString(),
+				Notes:    iprule.Notes.ValueString(),
+				Action:   iprule.Action.ValueString(),
+			})
+		}
+	}
+	return conf
+}
+
+func (r *firewallConfigResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+
+	var plan FirewallConfig
+	diags := req.Plan.Get(ctx, &plan)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	conf := plan.toClient()
+
+	out, err := r.client.PutFirewallConfig(ctx, conf)
+	if err != nil {
+		diags.AddError("failed to create firewall config", err.Error())
+	}
+
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	cfg := fromClient(out, plan)
+	diags = resp.State.Set(ctx, cfg)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r *firewallConfigResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var state FirewallConfig
+	diags := req.State.Get(ctx, &state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	out, err := r.client.GetFirewallConfig(ctx, state.ProjectID.ValueString(), state.TeamID.ValueString())
+	if err != nil {
```

Comment:
Do we need to handle 404s here?

Reply:
YES.  Actually ran into a couple bugs testing locally that made no sense.